### PR TITLE
Make broadcast recursion in `flatten` structural

### PR DESCRIPTION
**Pull request by Keno**
The inference enhancements in #29294 work quite well to prevent limiting
on many kinds of code. However, targetting TPUs, one code pattern it
struggeled with was a fairly large broadcast fusion in Flux:

     λ.(reshape(γ, affine_shape...) .* ((x .- μ) ./ σ) .+ reshape(β, affine_shape...))

The reason #29294 doesn't trigger is that the make_makeargs function used by the
implementation of Broadcast.flatten (which the TPU backend uses) had
a non-decreasing first argument (passing the return value of a previous
invocation of make_makeargs back in as the first argument). However,
that's not a fundamental limitation of the operation, but rather an
implementation choice. This PR switches that function's recursion pattern
to be purely structural, allowing inference to infer through it (with
the changes in #29294). As a result, ResNet50 infers properly.

### CODE REVIEW

File: base/broadcast.jl [hunk 1]
```diff
@@ -322,28 +322,48 @@ cat_nested(t::Broadcasted, rest...) = (cat_nested(t.args...)..., cat_nested(rest
...
-@inline function make_makeargs(makeargs, t::Tuple)
-    let makeargs = make_makeargs(makeargs, tail(t))
-        return @inline function(head, tail::Vararg{Any,N}) where N
-            (head, makeargs(tail...)...)
-        end
-    end
+"""
+    make_makeargs(makeargs_tail::Function, t::Tuple) -> Function
+
+Each element of `t` is one (consequtive) node in a broadcast tree.
```
Comment (nalimilan) [inline base/broadcast.jl]:
```suggestion
Each element of `t` is one (consecutive) node in a broadcast tree.
```

**Review by mbauman — APPROVED** [review]
Thank you for these comments!

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/29816#issuecomment-433500728) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/dc77c8dd27bf5ab0a4f0a05bc46ef049fc19832f/f538bcd_vs_010495f/report.md). cc @ararslan

**Comment by andreasnoack** [comment]
This PR seems to have broken Flux. cc @MikeInnes

**Pull request merged**